include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -1420,24 +1420,71 @@ class __brick_set_op
     }
 };
 
-// TODO: The implementation of shift left is reliant on exactly n (shift factor)
-// work items being launched by the parallel_for pattern, so it cannot be vectorized
-// or process multiple iterations per work items as is. For now, we must ensure that our
-// small submitter is launched in the SYCL backend's __parallel_for
 template <typename _ExecutionPolicy, typename _DiffType, typename _Range>
 struct __brick_shift_left
 {
-    using __base_t = walk_vector_or_scalar_base<_Range>;
+  private:
     using _ValueType = oneapi::dpl::__internal::__value_t<_Range>;
-    constexpr static bool __can_vectorize = false;
-    constexpr static std::uint16_t __preferred_vector_size = 1;
-    constexpr static std::uint16_t __preferred_iters_per_item = 1;
+    // Maximum size supported by compilers to generate vector instructions
+    constexpr static std::uint8_t __max_vector_size = 4;
+
+  public:
+    // We cannot process multiple iterations per work-item without leading to race conditions
+    constexpr static std::uint8_t __preferred_iters_per_item = 1;
+    constexpr static bool __can_vectorize =
+        oneapi::dpl::__ranges::__is_vectorizable_range<std::decay_t<_Range>>::value &&
+        std::is_fundamental_v<_ValueType> && sizeof(_ValueType) < 4;
+    constexpr static std::uint8_t __preferred_vector_size =
+        __can_vectorize ? oneapi::dpl::__internal::__dpl_ceiling_div(__max_vector_size, sizeof(_ValueType)) : 1;
+
     _DiffType __size;
     _DiffType __n;
 
     template <typename _IsFull, typename _ItemId>
     void
-    __scalar_path_impl(_IsFull __is_full, const _ItemId __idx, _Range __rng) const
+    __vector_path_impl(_IsFull __is_full, const _ItemId __idx, _Range __rng) const
+    {
+        const std::size_t __unsigned_size = __size;
+        const _DiffType __i = __idx - __n;
+        for (_DiffType __k = __n; __k < __size; __k += __n)
+        {
+            const _DiffType __read_offset = __k + __idx;
+            const _DiffType __write_offset = __k + __i;
+            if constexpr (_IsFull::value)
+            {
+                if (__k + __idx + __preferred_vector_size <= __size)
```

Comment:
This is a comment for all three implementations of this for loop body (scalar, vector full, and vector partial):

I believe we can calculate the number of iterations of the for outermost for loop (lets call that `M`), then run `M-1` iterations cleanly without any branch in the inner loop.  Finally, we can have the branch check in the inner loop for the last iteration, where we may interact with the end of the actual sequence.

Reply:
Yep, I agree we can definitely do this. I had avoided it since the original scalar path did not do this, and I was afraid of rushing anymore changes into this path at this point. I will take a look though.

Reply:
Filed here for now: https://github.com/uxlfoundation/oneDPL/issues/2021 as I do not want to rush the change and it would involve touching the initial scalar implementation.